src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,395 @@
+            /// </summary>
+            private int _numUserWaits = 0;
+            /// <summary>
+            /// The current calculated timeout of this wait thread.
+            /// </summary>
+            private int _currentTimeout = Timeout.Infinite;
+            /// <summary>
+            /// A lock for editing any handle registration (i.e. the fields above).
+            /// </summary>
+            private readonly LowLevelLock _registeredHandlesLock = new LowLevelLock();
+
+            /// <summary>
+            /// A list of removals of wait handles that are waiting for the wait thread to process.
+            /// </summary>
+            private readonly RegisteredWaitHandle[] _pendingRemoves = new RegisteredWaitHandle[WaitHandle.MaxWaitHandles - 1];
+            /// <summary>
+            /// The number of pending removals.
+            /// </summary>
+            private int _numPendingRemoves = 0;
+            /// <summary>
+            /// A lock for modifying the pending removals.
+            /// </summary>
+            private readonly LowLevelLock _removesLock = new LowLevelLock();
+
+            /// <summary>
+            /// An event to notify the wait thread that there are pending adds or removals of wait handles so it needs to wake up.
+            /// </summary>
+            private readonly AutoResetEvent _changeHandlesEvent = new AutoResetEvent(false);
+
+            /// <summary>
+            /// An event that is signaled when it is safe to dispose any handles that were queued to be removed previously.
+            /// </summary>
+            private readonly AutoResetEvent _safeToDisposeHandleEvent = new AutoResetEvent(false);
+
+            /// <summary>
+            /// Whether or not the wait thread has started.
+            /// </summary>
+            private bool _waitThreadStarted = false;
+            /// <summary>
+            /// A lock for the <see cref="_waitThreadStarted"/>.
+            /// </summary>
+            private LowLevelLock _waitThreadStartedLock = new LowLevelLock();
+
+            /// <summary>
+            /// The main routine for the wait thread.
+            /// </summary>
+            private void WaitThreadStart()
+            {
+                while (true)
+                {
+                    // We are going to be modifying the removals array and waiting on the remaining wait handles, so client code should wait
+                    // until we are not directly using the wait handles to allow user code to dispose of the WaitHandle.
+                    _safeToDisposeHandleEvent.Reset();
+                    ProcessRemovals();
+                    int signaledHandleIndex = WaitHandle.WaitAny(_waitHandles, _numUserWaits + 1, _currentTimeout);
+                    WaitHandle signaledHandle = signaledHandleIndex != WaitHandle.WaitTimeout ? _waitHandles[signaledHandleIndex] : null;
+                    ProcessRemovals();
```

Comment:
If the wait completed because some sync object got signaled, it cannot be removed first. The callback has to be raised to acknowledge the signal and then it can be removed.